game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java

```diff
@@ -0,0 +1,143 @@
+    targetUnits.retainAll(aliveUnits);
+  }
+
+  /**
+   * Splits up the firingUnits by suicideOnHit status and groups them by unit type
+   *
+   * @param name Name of the firing units
+   * @param firingUnits Collection of units that are firing
+   * @param targetUnits Collection of units that are being hit
+   * @return List of FiringGroup
+   */
+  public static List<FiringGroup> groupBySuicideOnHit(
+      final String name, final Collection<Unit> firingUnits, final Collection<Unit> targetUnits) {
+    final List<FiringGroup> groups = new ArrayList<>();
+    final List<Collection<Unit>> separatedBySuicide = separateSuicideOnHit(firingUnits);
+
+    // ensure each firing group has a unique name by adding prefixes
+    // if the firing groups have different types of suicide units
+    if (separatedBySuicide.size() == 1) {
+      groups.add(
+          new FiringGroup(
+              name,
+              name,
+              separatedBySuicide.get(0),
+              targetUnits,
+              Matches.unitIsSuicideOnHit().test(separatedBySuicide.get(0).iterator().next())));
+
+    } else if (separatedBySuicide.size() == 2
+        && Matches.unitIsSuicideOnHit().test(separatedBySuicide.get(0).iterator().next())
+            != Matches.unitIsSuicideOnHit().test(separatedBySuicide.get(1).iterator().next())) {
+      groups.addAll(
+          generateFiringGroupsWithOneSuicideAndOneNonSuicide(
+              name, targetUnits, separatedBySuicide));
+
+    } else {
+      groups.addAll(generateFiringGroups(name, targetUnits, separatedBySuicide));
+    }
+    return groups;
+  }
+
+  /**
+   * Separate the suicide on hit units from the others and group them by their type. The suicide on
+   * hit units need to fire separately so that they can be removed if they hit.
+   */
+  private static List<Collection<Unit>> separateSuicideOnHit(final Collection<Unit> units) {
+
+    final Map<UnitType, Collection<Unit>> map = new HashMap<>();
+    for (final Unit unit : CollectionUtils.getMatches(units, Matches.unitIsSuicideOnHit())) {
+      final UnitType type = unit.getType();
+      if (map.containsKey(type)) {
+        map.get(type).add(unit);
+      } else {
+        final Collection<Unit> unitList = new ArrayList<>();
+        unitList.add(unit);
+        map.put(type, unitList);
+      }
+    }
+
+    final List<Collection<Unit>> result = new ArrayList<>(map.values());
+    final Collection<Unit> remainingUnits =
+        CollectionUtils.getMatches(units, Matches.unitIsSuicideOnHit().negate());
+    if (!remainingUnits.isEmpty()) {
+      result.add(remainingUnits);
+    }
+    return result;
+  }
+
+  /** Handle the case where there are only two groups and one is suicide and the other is not */
+  private static List<FiringGroup> generateFiringGroupsWithOneSuicideAndOneNonSuicide(
+      final String name,
+      final Collection<Unit> targetUnits,
+      final List<Collection<Unit>> separatedBySuicide) {
```

Comment:
This param is interesting `List<Collection<Unit>>`. I wonder a bit what the ordering of the list is. If not ordered, then should it be just a `Collection`? That makes me wonder further if this would be better as a `Map<?, Collection<Unit>>`

For map key, if an enum would perhaps do well, something like `FiringGroupType { SUICIDE_ON_HIT, STANDARD }`

WDYT?

Reply:
If there were a map of firing group by type, then FiringGroup would not need the `suicideOnHit` boolean.

Reply:
Collection is fine.  The order doesn't matter.

A map won't work because there can be more than one suicideOnHit collection.  This specific method adds a simplified prefix for the special case where there is only one suicideOnHit collection and one non suicideOnHit collection (I tried to express that in the method name).  The logic to figure out which one is the suicideOnHit and which one is not the suicideOnHit is inside of this method.  If I move that logic out of this method, then this method becomes basically nothing and would be inlined.  But doing that will increase the complexity of the caller and codeclimate will complain.

Reply:
>  But doing that will increase the complexity of the caller and codeclimate will complain.

codeclimate is not perfect. There are times when the actual victory is from a fundamental change rather than a surface level. In other words, sometimes right right fix to codeclimate issues are not simple, surface level fixes. If a codeclimate warning is in the way of what is actually simpler, then codeclimate should be ignored in that instance.

Reply:
The encoding of the 'prefix' does strike me as complicated.

What are the different types of `suicideOnHit` collection? Perhaps we are talking instead a more complicated data structure, something like: `Map<UnitType, Map<SuicideOnHitFlag, Collection<Unit>>`. Generally, such a data structure is a candidate to be converted to a first class object. One would init it by passing a collection of units, perhaps calling a factory method, and the methods on the object would be queries to return units of a given type.  

I think there could be some advantages to that:
- readily tested
- extracts a good bit of detail
- readily mocked, could be useful for testing other logic that relies on units with certain characteristics

Reply:
I've made changes to how this works.  Take a look at the new code.